[motion-path] Implement offset-position: normal

nmoucht opened the pull request:
#### 958cc075a8612fc45d07e95dbfe9bad0547e3cd2
<pre>
[motion-path] Implement offset-position: normal
<a href="https://bugs.webkit.org/show_bug.cgi?id=259988">https://bugs.webkit.org/show_bug.cgi?id=259988</a>
rdar://113644765

Reviewed by Tim Nguyen.

Have the inital value of offset-position default to using normal rather than auto. Currently
implemented by adding a normal keyword to length and specifying this default value in
RenderStyle::initialOffsetPosition. We should update normalPositionForOffsetPath in future
patches to ensure the normal value is correct for path operations other than ray().

* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-position-interpolation.html:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/inheritance.html:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::create):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::isNormal):
(WebCore::valueForPositionOrAuto):
(WebCore::valueForOffsetShorthand):
* Source/WebCore/css/calc/CSSCalcValue.cpp:
(WebCore::createCSS):
* Source/WebCore/platform/Length.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/Length.h:
(WebCore::Length::initialize):
(WebCore::Length::isNormal const):
* Source/WebCore/platform/LengthFunctions.cpp:
(WebCore::valueForLength):
(WebCore::floatValueForLength):
* Source/WebCore/platform/LengthFunctions.h:
(WebCore::minimumValueForLength):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computeReplacedLogicalWidthUsing const):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialOffsetPosition):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPositionOrAuto):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;Length&gt;::encode):
(IPC::ArgumentCoder&lt;Length&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/267657@main">https://commits.webkit.org/267657@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/070d3c53362ccfc01b7b45f00b64bc82c794b6e7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17606 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18110 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19069 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16171 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17477 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20881 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17749 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18357 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17485 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17828 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15024 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19886 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15068 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15714 "8 failures") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16067 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15882 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20210 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16464 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13973 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15613 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19983 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2114 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16299 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1323,6 +1323,10 @@ bool isInitialValueForLonghand(CSSPropertyID longhand, const CSSValue& value)
+    case CSSPropertyOffsetPosition:
+        if (isValueID(value, CSSValueNormal))
+            return true;
+        break;
```
Could you change `initialValueForLonghand` instead of changing this? Notably this needs to be changed: https://searchfox.org/wubkat/rev/f61c5111898cf9fd9ee106240f085c69776dc8a8/Source/WebCore/css/parser/CSSPropertyParser.cpp#904

darinadler replied:
Definitely should do that. We should probably add a comment here saying that this function must _only_ contain cases where `initialValueForLonghand` can’t work correctly either because there are multiple ways to express the initial value internally or the type isn’t a number or a keyword.

Might want to look into adding more test coverage since `initialValueTextForLonghand` will return `"auto"` and not `"normal"` and `initialValueIDForLonghand` will return `CSSValueAuto` instead of `CSSValueNormal` and that might cause one or more web exposed bugs.

nmoucht replied:
Right fixed. We were actually failing a test because of this but it wasn't obvious since we were failing it previously as well. So maybe the coverage is ok here.

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -38,6 +38,13 @@ static FloatPoint offsetFromContainer(const RenderObject& renderer, RenderBlock&
     return FloatPoint(FloatPoint(offsetFromContainingBlock) - referenceRect.location());
 }
 
+static FloatPoint getNormalPositionForOffsetPath(PathOperation* operation, const FloatRect& referenceRect)
```
```suggestion
static FloatPoint normalPositionForOffsetPath(PathOperation* operation, const FloatRect& referenceRect)
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/platform/Length.h:
```diff
@@ -37,6 +37,7 @@ namespace WebCore {
 
 enum class LengthType : uint8_t {
     Auto,
+    Normal,
```
@smfr @alanbaradlay @anttijk @darinadler Do you know if it's OK to add this new LengthType?

darinadler replied:
It should be OK to add a type like this if it’s only used in certain places. However, we need to be cautious about how this new type of value could flow into code that isn’t prepared for it. As far as I can tell this patch gets that right, but in the places where we use `if` statements instead of `switch` statements or use `switch` statements with a `default`, there’s no guarantee we found all the ones that we need to modify.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1023,13 +1023,16 @@ class LengthPointOrAutoPropertyWrapper : public LengthPointPropertyWrapper {
+        bool fromIsNormal = value(from).x().isNormal() && value(from).y().isNormal();
+        bool toIsNormal = value(to).x().isNormal() && value(to).y().isNormal();
+
+        return (!fromIsAuto && !toIsAuto) && (!fromIsNormal && !toIsNormal);
```
It’s wasteful to call `value` 4 times each on `from` and `to` if it isn’t guaranteed to be inlined and common sub expressions eliminated. I suggest calling it only once on each and using some local variables. Also there’s no need for the parentheses here, which give the impression of a grouping logic that does not actually exist.

But also since we don’t support a mix of `auto` and `normal` and other length types in `x` and `y` axes there is no need to check `y` at all. The code would malfunction if we had a mix, returning `true` in a case where we cannot interpolate.

darinadler commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -512,6 +512,8 @@ Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(String value)
 Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(const Length& length)
 {
     switch (length.type()) {
+    case LengthType::Normal:
```
Not sure why you added this at the top. If I was editing I would probably make them alphabetical since the list is so long.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -545,6 +547,7 @@ Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(const Length& length, const Ren
 {
     switch (length.type()) {
     case LengthType::Auto:
+    case LengthType::Normal:
```
Here we are adding just after `Auto`. Again, I think alphabetical might be good.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1638,10 +1638,17 @@ static bool isAuto(const LengthPoint& position)
-    if (position.x().isAuto() && position.y().isAuto())
+    if (isAuto(position))
         return CSSPrimitiveValue::create(CSSValueAuto);
+    if (isNormal(position))
```
This function is named `valueForPositionOrAuto`. Adding _normal_ to it makes the function name **incorrect**. We need to change the function name, either to be more abstract, or to include the word _normal_.

nmoucht replied:
Changed to `valueForPositionOrAutoOrNormal`.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2719,7 +2726,7 @@ static Ref<CSSValue> valueForOffsetShorthand(const RenderStyle& style)
 
-    if (!isAuto(style.offsetPosition()))
+    if (!isAuto(style.offsetPosition()) && !isNormal(style.offsetPosition()))
         innerList.append(valueForPositionOrAuto(style, style.offsetPosition()));
```
This should call `valueForPosition` since the code already checks for _auto_.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -332,6 +332,7 @@ static_assert(sizeof(Length) == sizeof(SameSizeAsLength), "length should stay sm
 static TextStream& operator<<(TextStream& ts, LengthType type)
 {
     switch (type) {
+    case LengthType::Normal: ts << "normal"; break;
```
Again not sure how we are ordering things. I tend to lean on alphabetical order in long lists like these.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -352,6 +353,7 @@ static TextStream& operator<<(TextStream& ts, LengthType type)
 TextStream& operator<<(TextStream& ts, Length length)
 {
     switch (length.type()) {
+    case LengthType::Normal:
```
Again not sure how we are ordering things. I tend to lean on alphabetical order in long lists like these.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/platform/Length.h:
```diff
@@ -92,6 +93,7 @@ struct Length {
 
     LengthType type() const;
 
+    bool isNormal() const;
```
Same issue about sorting things alphabetically that I mentioned elsewhere. i’m going to stop commenting about it in each case but I think it’s nice to be thoughtful about the sequence and if there’s no super obvious logical order then alphabetical is better than arbitrary order.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -38,6 +38,13 @@ static FloatPoint offsetFromContainer(const RenderObject& renderer, RenderBlock&
+{
+    if (is<RayPathOperation>(operation))
+        return referenceRect.center();
+    return { 0.0, 0.0 };
```
Should be able to write just `{ }` unless you have reason to think there is greater clarity by explicitly specifying `0.0`.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -47,6 +54,9 @@ std::optional<MotionPathData> MotionPath::motionPathDataForRenderer(const Render
     auto startingPositionForOffsetPosition = [&](const LengthPoint& offsetPosition, const FloatRect& referenceRect, RenderBlock& container) -> FloatPoint {
         // FIXME: Implement offset-position: normal.
+        // If offset-position is normal, the element does not have an offset position
+        if (offsetPosition.x().isNormal() && offsetPosition.y().isNormal())
```
Why does this check both `x` and `y`? Can one be normal without the other also being normal? I see that the existing code does this for auto, but I don’t think it’s helpful in that case either. Seems clear it should just check one axis. I suggest `x`.

nmoucht replied:
Right changed to just check x axis.

darinadler commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -415,7 +415,7 @@ inline LengthPoint RenderStyle::initialObjectPosition() { return { { 50.0f, Leng
 inline LengthPoint RenderStyle::initialOffsetAnchor() { return { }; }
 inline Length RenderStyle::initialOffsetDistance() { return zeroLength(); }
-inline LengthPoint RenderStyle::initialOffsetPosition() { return { }; }
+inline LengthPoint RenderStyle::initialOffsetPosition() { return LengthPoint({ LengthType::Normal }, { LengthType::Normal }); }
```
Should be able to use braces without types like this `{ { LengthType::Normal }, { LengthType::Normal } }`. Could even consider letting the `y` axis just be the default value if we fix code to understand the special value comes from the `x` axis.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -483,6 +483,8 @@ inline LengthPoint BuilderConverter::convertPositionOrAuto(BuilderState& builder
     if (value.isPair())
         return convertPosition(builderState, value);
+    if (value.valueID() == CSSValueNormal)
+        return LengthPoint({ LengthType::Normal }, { LengthType::Normal });
```
Should be able to use braces without types like this `{ { LengthType::Normal }, { LengthType::Normal } }` or even find a way to share code with `RenderStyle::initialOffsetPosition` to avoid repeating a long expression.

nmoucht replied:
Fixed.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1638,10 +1638,17 @@ static bool isAuto(const LengthPoint& position)
 
+static bool isNormal(const LengthPoint& position)
+{
+    return position.x().isNormal() && position.y().isNormal();
```
I am puzzled why the `isAuto` function and this new function check both `x` and `y`. Clearly the code does not correctly support a mix of `normal` or `auto` and other types of length? Perhaps these should check one axis and assert the other. Or just check one axis.

nmoucht replied:
Changed to check x axis.

darinadler commented on Source/WebCore/platform/Length.cpp:
```diff
@@ -284,7 +284,7 @@ static Length blendMixedTypes(const Length& from, const Length& to, const Blendi
 Length blend(const Length& from, const Length& to, const BlendingContext& context)
 {
-    if ((from.isAuto() || to.isAuto()) || (from.isUndefined() || to.isUndefined()))
+    if ((from.isAuto() || to.isAuto()) || (from.isUndefined() || to.isUndefined()) || (from.isNormal() || to.isNormal()))
```
I suggest removing the parentheses that suggest a grouping that does not exist.

nmoucht replied:
Fixed.

darinadler reviewed (changes requested):
Looks great!

It’s important to fix the default value and figure out why doing it wrong didn’t affect any test results. We have insufficient test coverage for this default value, it seems.

darinadler commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -47,6 +54,9 @@ std::optional<MotionPathData> MotionPath::motionPathDataForRenderer(const Render
 
     auto startingPositionForOffsetPosition = [&](const LengthPoint& offsetPosition, const FloatRect& referenceRect, RenderBlock& container) -> FloatPoint {
         // FIXME: Implement offset-position: normal.
```
Please update this comment since the patch implements at least part of this feature. The comment is confusing right next to code that does seem to implement it.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1023,13 +1023,13 @@ class LengthPointOrAutoPropertyWrapper : public LengthPointPropertyWrapper {
+        auto valueTo = value(to);
 
-        return (!fromIsAuto && !toIsAuto); 
+        return (!valueFrom.x().isAuto() && !valueTo.x().isAuto() && !valueFrom.x().isNormal() && !valueTo.x().isNormal());
```
nit: you can remove the brackets around the return value

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -54,7 +61,9 @@ std::optional<MotionPathData> MotionPath::motionPathDataForRenderer(const Render
 
     auto startingPositionForOffsetPosition = [&](const LengthPoint& offsetPosition, const FloatRect& referenceRect, RenderBlock& container) -> FloatPoint {
-        // FIXME: Implement offset-position: normal.
+        // If offset-position is normal, the element does not have an offset position
```
```suggestion
        // If offset-position is normal, the element does not have an offset starting position.
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -483,6 +483,8 @@ inline LengthPoint BuilderConverter::convertPositionOrAuto(BuilderState& builder
 {
```
Should this function be called `convertPositionOrAutoOrNormal` ?

nmoucht replied:
Fixed.

nmoucht committed:
[motion-path] Implement offset-position: normal
https://bugs.webkit.org/show_bug.cgi?id=259988
rdar://113644765

Reviewed by Tim Nguyen.

Have the inital value of offset-position default to using normal rather than auto. Currently
implemented by adding a normal keyword to length and specifying this default value in
RenderStyle::initialOffsetPosition. We should update normalPositionForOffsetPath in future
patches to ensure the normal value is correct for path operations other than ray().

* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-position-interpolation.html:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/inheritance.html:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::CSSPrimitiveValue::create):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::isNormal):
(WebCore::valueForPositionOrAuto):
(WebCore::valueForOffsetShorthand):
* Source/WebCore/css/calc/CSSCalcValue.cpp:
(WebCore::createCSS):
* Source/WebCore/platform/Length.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/Length.h:
(WebCore::Length::initialize):
(WebCore::Length::isNormal const):
* Source/WebCore/platform/LengthFunctions.cpp:
(WebCore::valueForLength):
(WebCore::floatValueForLength):
* Source/WebCore/platform/LengthFunctions.h:
(WebCore::minimumValueForLength):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computeReplacedLogicalWidthUsing const):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialOffsetPosition):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPositionOrAuto):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<Length>::encode):
(IPC::ArgumentCoder<Length>::decode):

Canonical link: https://commits.webkit.org/267657@main
